Samples/DirectMLNpuInference/main.cpp

```diff
@@ -34,32 +59,49 @@ void InitializeDirectML(ID3D12Device1** d3dDeviceOut, ID3D12CommandQueue** comma
     ComPtr<IDXCoreAdapter> adapter;
     if (factory)
     {
-        const GUID dxGUIDs[] = { DXCORE_ADAPTER_ATTRIBUTE_D3D12_CORE_COMPUTE };
-        ComPtr<IDXCoreAdapterList> adapterList;
-        THROW_IF_FAILED(factory->CreateAdapterList(ARRAYSIZE(dxGUIDs), dxGUIDs, IID_PPV_ARGS(&adapterList)));
-        for (uint32_t i = 0, adapterCount = adapterList->GetAdapterCount(); i < adapterCount; i++)
+        for (auto& allowedGuid : dxGuidAllowedAttributes)
         {
-            ComPtr<IDXCoreAdapter> currentGpuAdapter;
-            THROW_IF_FAILED(adapterList->GetAdapter(static_cast<uint32_t>(i), IID_PPV_ARGS(&currentGpuAdapter)));
-
-            if (!forceComputeOnlyDevice && !forceGenericMLDevice)
-            {
-                // No device restrictions
-                adapter = std::move(currentGpuAdapter);
-                break;
-            }
-            else if (forceComputeOnlyDevice && currentGpuAdapter->IsAttributeSupported(DXCORE_ADAPTER_ATTRIBUTE_D3D12_CORE_COMPUTE))
-            {
-                adapter = std::move(currentGpuAdapter);
-                break;
-            }
-            else if (forceGenericMLDevice && currentGpuAdapter->IsAttributeSupported(DXCORE_ADAPTER_ATTRIBUTE_D3D12_GENERIC_ML))
+            if (adapter != nullptr) continue;
+            
+            ComPtr<IDXCoreAdapterList> adapterList;
+            THROW_IF_FAILED(factory->CreateAdapterList(1, &allowedGuid, IID_PPV_ARGS(&adapterList)));
+                
+            for (uint32_t i = 0, adapterCount = adapterList->GetAdapterCount(); i < adapterCount; i++)
             {
+                ComPtr<IDXCoreAdapter> currentGpuAdapter;
+                THROW_IF_FAILED(adapterList->GetAdapter(static_cast<uint32_t>(i), IID_PPV_ARGS(&currentGpuAdapter)));
+
+                bool isAdapterValid = true;
+
+                // filter out adapters with disallowed attributes
+                for (auto& disallowedGuid : dxGuidDisallowedAttributes)
+                {
+                    if (currentGpuAdapter->IsAttributeSupported(disallowedGuid)) { isAdapterValid = false; }
+                }
+
+                // filter out adapters that doesn match all required attributes
+                for (auto& requiredGuid : dxGuidRequiredAttributes)
+                {
+                    if (!currentGpuAdapter->IsAttributeSupported(requiredGuid)) { isAdapterValid = false; }
+                }
+
+                if (!isAdapterValid) { continue; }
+
                 adapter = std::move(currentGpuAdapter);
                 break;
             }
         }
     }
+
+    if (adapter)
+    {
+        size_t adapterNameSize = 0;
+        THROW_IF_FAILED(adapter->GetPropertySize(DXCoreAdapterProperty::DriverDescription, &adapterNameSize));
+        char* adapterName = (char*)malloc(adapterNameSize);
```

Comment:
Where is the memory released?
Better to just use std::string.

Reply:
https://github.com/microsoft/DirectML/blob/a87891ebb1ff7cb3ba142919eda54e3a803ab597/DxDispatch/src/dxdispatch/Adapter.h#L32-L51